### PR TITLE
Quests System

### PR DESCRIPTION
**Pull request by rafd**
This PR adds a quest system to Braid. The purpose of the quest system is to introduce new users to various features.

screenshot 2016-08-16 21 17 43

Each quest also includes a short video (GIF) of how to perform the action:

screenshot 2016-08-18 19 24 21

I needed to refactor a few parts of the codebase to make quests work.

Also, I put all quest related code in its own folders (one on the client, and one on the server) rather than spreading it across existing files (ie. instead of adding quest-related subscriptions to `braid.client.state` they are in `briad.client.quests.subscriptions`). My hope is that other parts of the codebase could be organized like this in the future (where it makes sense; I foresee a large amount under a 'core' or 'common' ns, b/c many features are interdependent)

Currently, only 4 quests are implemented, with many more planned.

### CODE REVIEW

File: src/braid/client/quests/helpers.cljs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+       (filter (fn [quest-record]
+                 (= (quest-record :quest-record/state) :active)))
+       (sort-by :quest/order)))
+
+(defn get-next-quest [state]
+  (let [quest-ids-with-records (->> state
+                                    :quest-records
+                                    vals
+                                    (map (fn [quest-record]
+                                           (quest-record :quest-record/quest-id)))
```
Comment (jamesnvc) [inline src/braid/client/quests/helpers.cljs]: Could be just `(map :quest-record/quest-id)`?

File: src/braid/client/dispatcher.cljs [hunk 2]
```diff
@@ -1,10 +1,13 @@
...
+            [braid.client.state.handler.core :refer [handler]]
+            [braid.client.quests.handler :refer [quests-handler]]))
 
 (defn dispatch!
   ([event args]
    (println event)
-   (store/transact! (handler @store/app-state [event args])))
+   (store/transact! (-> @store/app-state
+                        (quests-handler [event args])
+                        (handler [event args]))))
```
Comment (jamesnvc) [inline src/braid/client/dispatcher.cljs]: I feel like I'll need to think about this for a bit...I get why it's like this, but it makes me a little leery that it needs to tie so deeply into such a fundamental part of the system.  Should there be a more general way for some subsystem to say "tell me about these sorts of events" or something?

File: src/braid/client/state.cljs [hunk 3]
```diff
@@ -1,15 +1,11 @@
 (ns braid.client.state
   (:require [reagent.ratom :include-macros true :refer-macros [reaction]]
             [braid.client.store :as store]
-            [clojure.set :refer [union intersection subset?]])
+            [clojure.set :refer [union intersection subset?]]
+[braid.client.state.subscription :refer [subscription]]
```
Comment (jamesnvc) [inline src/braid/client/state.cljs]: identation

File: src/braid/server/db/user.clj [hunk 4]
```diff
@@ -12,13 +13,15 @@
...
-       (create-entity! conn)
-       db->user))
+  (let [user (->> {:user/id id
+                   :user/email email
+                   :user/avatar avatar
+                   :user/nickname (or nickname (-> email (string/split #"@") first))
+                   :user/password-token (password/encrypt password)}
+                  (create-entity! conn)
+                  db->user)
+        _ (activate-first-quests! conn id)]
```
Comment (jamesnvc) [inline src/braid/server/db/user.clj]: Could just move the activate out of the `let` - `(let [user ...] (activate-first-quests! ...) user)`

**Comment by jamesnvc** [comment]
Looks good; my two concerns are as follows:

1. I fear it isn't obvious that you need to do the action multiple times - when I first did the "start a conversation" thing and got one checkmark, I assumed the next step was some other more complicated thing, not do just do the same thing again

2. As mentioned in the diff, I am a little leery of how quests need to add a handler into `dispatch!`.  I really like how orthogonal the rest of the quest stuff is (living in its own sub-ns) so I wonder if there's a better way to have things observe changes to the state without having to keep threading more handler functions in there.  This is obviously more of a nitpicky thing than a concrete concern and could probably wait until we start building other things that also want to observe state and try to factor that notion out then, but I think it's worth pondering.

File: src/braid/client/dispatcher.cljs (hunk 2 above)
Comment (rafd) [inline src/braid/client/dispatcher.cljs]: We could extend the dispatching system to allow for multiple listeners per event, in lieu of defmethod.

Comment (jamesnvc) [inline src/braid/client/dispatcher.cljs]: What if `dispatch` put events in a queue (e.g. a core.async chan) that then interested parties could subscribe to - the primary one would be the store, which would call handlers, but then other things could just add a tap?

**Comment by rafd** [comment]
Update to display number of times an action must be performed:

screenshot 2016-08-18 21 43 34

**Comment by jamesnvc** [comment]
Btw, for gif recording, there's a very useful utility for Linux called [byzanz](http://linux.die.net/man/1/byzanz-record) that may be worth mentioning in the script as an option for non-OS X

**Comment by rafd** [comment]
Let's merge this in and add a task to extract it from `dispatch!`? 
I don't think it will become a habit of ours to to keep adding to `dispatch!`

**Comment by jamesnvc** [comment]
Sure, sounds good

**Pull request merged**